lib/cartopy/mpl/gridliner.py

```diff
@@ -327,58 +338,257 @@ def _draw_gridliner(self, nx=None, ny=None, background_patch=None):
-                if self.xlabels_top:
-                    self._add_gridline_label(x, axis='x', upper_end=True)
-
-        if self.ylabels_left or self.ylabels_right:
-            self._assert_can_draw_ticks()
-            self.yformatter.set_locs(y_label_points)
-            for y in y_label_points:
-                if self.ylabels_left:
-                    self._add_gridline_label(y, axis='y', upper_end=False)
-                if self.ylabels_right:
-                    self._add_gridline_label(y, axis='y', upper_end=True)
+        """
+        if renderer is None or not self._labels:
+            return
+        paths = []
+        outline_path = None
+        delta_angle = 22.5
+        max_delta_angle = 45
+        axes_children = self.axes.get_children()
+        for priority, artist in self._labels:
+
+            if artist not in axes_children:
+                warn('The labels if this gridliner do not belongs'
+                     'to the gridliner axes')
+
+            # Compute angles to try
+            orig_specs = {'rotation': artist.get_rotation(),
+                          'ha': artist.get_ha(),
+                          'va': artist.get_va()}
+            angles = [None]
+            for abs_delta_angle in np.arange(delta_angle, max_delta_angle+1,
+                                             delta_angle):
+                for sign_delta_angle in (1, -1):
+                    angle = artist._angle + sign_delta_angle * abs_delta_angle
+                    angles.append(angle)
+
+            # Loop on angles until it works
+            for angle in angles:
+
+                if angle is not None:
+                    specs, _ = self._segment_angle_to_text_specs(angle)
+                    plt.setp(artist, **specs)
+
+                artist.update_bbox_position_size(renderer)
+                this_patch = artist.get_bbox_patch()
+                this_path = this_patch.get_path().transformed(
+                        this_patch.get_transform())
+                visible = False
+                for path in paths:
+
+                    # Check it does not overlap another label
+                    if this_path.intersects_path(path):
+                        break
+
+                else:
+
+                    # Finally check that it does not overlap the map
+                    if outline_path is None:
+                        outline_path = self.axes.background_patch.get_path(
+                            ).transformed(self.axes.transData)
+                    visible = not outline_path.intersects_path(this_path)
+
+                    # Good
+                    if visible:
+                        break
+
+            # Action
+            artist.set_visible(visible)
+            if not visible:
+                plt.setp(artist, **orig_specs)
+            else:
+                paths.append(this_path)
+                # is this necessary?
+#                artist.update_bbox_position_size(renderer)
```

Comment:
I take it this is left in here because you are not sure whether it makes a difference or not.  I am happy to leave it in here unless you have any objections?